Add "cache" source support.

yoshisatoyanagisawa opened the pull request:
The ServiceWorker static routing API allows to look up a response from the cache storage directly upon the rule.  This is a specification update to support that.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/4.html" title="Last updated on Jan 26, 2024, 11:00 AM UTC (c6628d1)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/4/9709650...c6628d1.html" title="Last updated on Jan 26, 2024, 11:00 AM UTC (c6628d1)">Diff</a>

yoshisatoyanagisawa committed:
Add "cache" source support.

yoshisatoyanagisawa commented:
@sisidovski @domenic Can I ask you to review the update?

In https://github.com/yoshisatoyanagisawa/ServiceWorker/pull/3, @domenic suggested to omit the condition.  However, I am hesitate to add several steps to the top level of the Handle Fetch algorithm (more top level variables if people implement this literally), and I kept on having the condition.  Let me know what do you think about that.

Thank you,
Yoshisato.

sisidovski commented on docs/index.bs:
```diff
@@ -3110,7 +3110,15 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
-          1. If running [=Get Router Source=] algorithm with |registration|'s <a>active worker</a> and |request| returns "network", return null.
+          1. Let |source| be a result of running [=Get Router Source=] algorithm with |registration|'s <a>active worker</a> and |request|.
+          1. If |source| is "network", return null.
+          1. Else if |source| is "cache":
```
`Else if |source| is "cache", then:`?

yoshisatoyanagisawa replied:
done.

sisidovski commented on docs/index.bs:
```diff
@@ -3110,7 +3110,15 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
       1. Else if |registration|'s <a>active worker</a>'s [=service worker/list of router rules=] is [=list/is not empty=]:
-          1. If running [=Get Router Source=] algorithm with |registration|'s <a>active worker</a> and |request| returns "network", return null.
+          1. Let |source| be a result of running [=Get Router Source=] algorithm with |registration|'s <a>active worker</a> and |request|.
+          1. If |source| is "network", return null.
```
"network" can be `{{RouterSource/"network"}}`?

yoshisatoyanagisawa replied:
done.

sisidovski reviewed (commented):
Thanks! Direction looks good to me.

yoshisatoyanagisawa committed:
Addressed the comment.

- added "then".
- Changed source to {{RouterSource/"..."}}.

sisidovski reviewed (approved):
LGTM

yoshisatoyanagisawa commented:
Thanks for the review.
@domenic Can I ask you to take a look?

domenic reviewed (changes requested):
This unfortunately is pretty broken in a few ways, due to how it's attempting to use the public API `Cache/matchAll`.

- `Cache/matchAll()` operates on `Request` and `Response` objects, but you want the "response" concept.
- `Cache/matchAll()` can only be called from the main thread, but this is happening on a background thread.
- `Cache/matchAll()` can return multiple responses. You seem to be assuming only a single response.

I think you need to either copy or abstract out all the steps of matchAll which are applicable into a separate algorithm, which operates in a background thread and doesn't involve JavaScript objects. E.g., you can skip steps 1-2 and just let _r_ be the "response" you're passing in. You mostly only need steps 5.3-4, it looks like.

yoshisatoyanagisawa committed:
Merge branch 'static_routing_api' into yy_workspace

yoshisatoyanagisawa committed:
Use [=Query Cache=] directly to get Response in cache.

yoshisatoyanagisawa committed:
Add COEP support.

yoshisatoyanagisawa commented:
@domenic Thanks for the comment.  I tried to stop using `cache.matchAll()`, and directly use the essential part of steps I need for getting the Response.  Please take another look.

domenic commented on docs/index.bs:
```diff
@@ -3112,7 +3112,19 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
           1. Return null.
       1. Else if |registration|'s <a>active worker</a>'s [=service worker/list of router rules=] is [=list/is not empty=]:
-          1. If running [=Get Router Source=] algorithm with |registration|'s <a>active worker</a> and |request| returns "network", return null.
+          1. Let |source| be a result of running [=Get Router Source=] algorithm with |registration|'s <a>active worker</a> and |request|.
```
```suggestion
          1. Let |source| be the result of running [=Get Router Source=] algorithm with |registration|'s <a>active worker</a> and |request|.
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3112,7 +3112,19 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. If |requestResponses| is an empty [=list=], return null.
+              1. Else:
+                  1. Let |requestResponse| be the first element of |requestResponses|.
+                  1. Let |response| be a copy of |requestResponse|'s response.
```
Is "a copy of" necessary? It doesn't seem well-defined. But I see it is used elsewhere, so maybe it is fine.

yoshisatoyanagisawa replied:
I was inspired by "Add a copy of requestResponse’s response to responses." in https://w3c.github.io/ServiceWorker/#cache-matchall.  However, I guess it is fine to omit "a copy of".  I have removed that.

domenic commented on docs/index.bs:
```diff
@@ -3112,7 +3112,19 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. Else:
+                  1. Let |requestResponse| be the first element of |requestResponses|.
+                  1. Let |response| be a copy of |requestResponse|'s response.
+                  1. Let |settingsObject| be a new [=environment settings object=].
```
Creating a new environment settings object doesn't seem correct. That represents a new Window or WorkerGlobalScope, but you haven't created one of those at the same time.

Since the purpose of this seems to be some sort of COOP check, shouldn't you be using the service worker's settings object?

Or... hmm, is the problem here that the ServiceWorkerGlobalScope might not be created at all yet? That seems hard to work around. So many algorithms, e.g. https://fetch.spec.whatwg.org/#cross-origin-resource-policy-check , seem to assume that you'll have the actual environment settings object. E.g. https://fetch.spec.whatwg.org/#queue-a-cross-origin-embedder-policy-corp-violation-report assumes that a global object exists.

How does this work in the implementation? Do we still send reports somehow?

yoshisatoyanagisawa replied:
It is actually implemented like: https://source.chromium.org/chromium/chromium/src/+/main:content/browser/service_worker/service_worker_cache_storage_matcher.cc;l=49;drc=c686e8f4fd379312469fe018f5c390e9c8f20d0d
It gets COEP from `ServiceWorkerVersion`.  Since I did not have any ideas where to plumb the report, it actually do not.

domenic replied:
That seems problematic... so if you start using service worker static routing, you lose COOP reporting? I think some of our partners are using COOP reporting heavily.

yoshisatoyanagisawa replied:
Ah, I did not know that.  Since half of the code calling AddReceiver sets std::NullRemote() as COEP reporter, I did not thought that important.
https://source.chromium.org/chromium/chromium/src/+/main:components/services/storage/public/mojom/cache_storage_control.mojom;l=43;drc=3cc22c22d7637bc5604e8fef3b0882b51a762901

To be clear, for the cache lookup, COEP violation may not be reported.  Other cases won't be affected.

domenic replied:
Hmm. I think it is worth starting a discussion with the teams that own COEP maybe?

(Also, sorry for stating COOP. I guess we are only talking about COEP here, not COOP.)

yoshisatoyanagisawa replied:
Upon the suggestion from the COEP folk, I made request's COEP is used instead of SW's.  Will you take another look at this?
Thank you.

yoshisatoyanagisawa committed:
Addressed some of comments.

yoshisatoyanagisawa committed:
Avoid creating a new settings object.

yoshisatoyanagisawa committed:
Use |client| as settingsObject to evaluate COEP.

When the ServiceWorker static routing API is used and source other than
"fetch-event" is used, results are directly returned without running
ServiceWorker.  In that case, we use the request's COEP instead of
ServiceWorker's COEP.

yoshisatoyanagisawa commented:
By the way, I have no idea how I can resolve the conflict.  I tried to resolve it in https://github.com/yoshisatoyanagisawa/ServiceWorker/pull/4/commits/3a187d82735ef336da2886c780576fcdbd138055, but it seems not help.
Is it better to close this and open yet another PR with the same change?

sisidovski commented:
fyi, I don't see any conflict on this PR.

![image](https://github.com/yoshisatoyanagisawa/ServiceWorker/assets/1642255/8cf716d9-3c85-4c14-a73e-66df9dc4e6f4)

domenic reviewed (approved):
LGTM. I think this will indeed use the document's policy container, etc. to check and report COEP violations. Please be sure to write web platform tests for this case, e.g. maybe similar to https://github.com/web-platform-tests/wpt/blob/master/html/cross-origin-embedder-policy/cache-storage-reporting-service-worker.https.html

yoshisatoyanagisawa commented:
Thanks for the review.

yoshisatoyanagisawa commented:
Thanks @sisidovski to write COEP to a list of missing WPTs.

yoshisatoyanagisawa commented:
(sorry, the list is managed internally.)
